File: src-tauri/src/project.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

use crate::language::{Language, LanguagePair};
use crate::transcript::{split_label, Segment, SpeakerNames, SrtContent, Transcript};

pub mod commands;
mod current;
mod files;
pub mod glossary;
mod history;
pub mod versions;

pub use current::{CurrentProject, ProjectView, ResourceView, RunningMode, SegmentSpan};
#[cfg(test)]
pub(crate) use files::HISTORY_DIR;
use glossary::TranslationGlossary;
use history::UndoHistory;

/// The opened directory: its Primary Language, the Language of its last translation,
/// its Resources, the Current Resource and the Translation Glossary once loaded.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Project {
    pub directory: PathBuf,
    pub language: Language,
    pub translation_language: Option<Language>,
    pub translation_glossary: Option<TranslationGlossary>,
    pub options: ProjectOptions,
    pub resources: Vec<Resource>,
    pub current: Option<CurrentResource>,
    /// The Undo History of each Resource changed since the Project was opened, by its name.
    pub undo_histories: HashMap<String, UndoHistory>,
}

/// The Resource the editor shows, with its Segments as read from the directory and edited since.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CurrentResource {
    pub name: String,
    pub transcript: Transcript,
    /// The Language of the translations its Segments carry.
    pub translation: Option<Language>,
    /// What its subtitle files held when Tsuzuri last read or wrote them, to tell a change made elsewhere.
    pub subtitle_digests: Vec<SubtitleDigest>,
}

/// A subtitle file and a digest of what it held, or `None` while it did not exist.
pub type SubtitleDigest = (PathBuf, Option<u64>);

/// Why the Project could not answer: it has no Resource by the name asked for, or none is current.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ProjectError {
    NoResource,
}

impl Project {
    /// The Languages a Translation Glossary's `source,target` header stands for: the Primary
    /// Language and the translation Language, once the Project has one.
    fn source_target(&self) -> Option<LanguagePair> {
        self.translation_language.map(|target| LanguagePair {
            source: self.language,
            target,
        })
    }

    /// The Primary Language and `translation` in the Bilingual Order.
    fn bilingual_languages(&self, translation: Option<Language>) -> [Option<Language>; 2] {
        match self.options.bilingual_order {
            BilingualOrder::OriginalFirst => [Some(self.language), translation],
            BilingualOrder::TranslationFirst => [translation, Some(self.language)],
        }
    }

    /// What the Translation Glossary calls each Speaker in `language`, by its name in the
    /// Primary Language; none without a glossary.
    fn speaker_names(&self, language: Option<Language>) -> HashMap<String, String> {
        let pair = language.map(|target| LanguagePair {
            source: self.language,
            target,
        });
        match (&self.translation_glossary, pair) {
            (Some(glossary), Some(pair)) => glossary.speaker_names(pair),
            _ => HashMap::new(),
        }
    }

    /// `transcript` as a Bilingual SRT in the Bilingual Order, its translation into `translation`.
    fn bilingual_srt(&self, transcript: &Transcript, translation: Option<Language>) -> String {
        let translated_names = self.speaker_names(translation);
        match self.options.bilingual_order {
            BilingualOrder::OriginalFirst => transcript.to_srt_with(
                SrtContent::Bilingual,
                &SpeakerNames {
                    translation: translated_names,
                    ..SpeakerNames::default()
                },
            ),
            BilingualOrder::TranslationFirst => translation_first(transcript).to_srt_with(
                SrtContent::Bilingual,
                &SpeakerNames {
                    text: translated_names,
                    ..SpeakerNames::default()
                },
            ),
        }
    }

    fn config(&self) -> ProjectConfig {
        ProjectConfig {
            language: Some(self.language),
            translation_language: self.translation_language,
            options: self.options.clone(),
        }
    }

    /// The Current Resource as SRT, a Bilingual SRT in the Bilingual Order.
    fn to_srt(&self, content: SrtContent) -> Result<String, ProjectError> {
        let current = self.current()?;
        let transcript = &current.transcript;
        Ok(match content {
            SrtContent::Bilingual => self.bilingual_srt(transcript, current.translation),
            SrtContent::Translation => transcript.to_srt_with(
                content,
                &SpeakerNames {
                    translation: self.speaker_names(current.translation),
                    ..SpeakerNames::default()
                },
            ),
            SrtContent::Original => transcript.to_srt(content),
        })
    }

    fn resource(&self, name: &str) -> Result<&Resource, ProjectError> {
        self.resources
            .iter()
            .find(|resource| resource.name == name)
            .ok_or(ProjectError::NoResource)
    }

    fn current(&self) -> Result<&CurrentResource, ProjectError> {
        self.current.as_ref().ok_or(ProjectError::NoResource)
    }

    fn current_mut(&mut self) -> Result<&mut CurrentResource, ProjectError> {
        self.current.as_mut().ok_or(ProjectError::NoResource)
    }
}

/// The translated Segments alone, each with its translation as its text.
fn translation_only(transcript: &Transcript) -> Transcript {
    Transcript {
        segments: transcript
            .segments
            .iter()
            .filter_map(|segment| {
                let translation = segment.translation.as_deref()?;
                (!translation.trim().is_empty()).then(|| Segment {
                    text: translation.to_string(),
                    translation: None,
                    ..segment.clone()
                })
            })
            .collect(),
    }
}

/// The translated Segments of `transcript` as the SRT of its translation, each Speaker called
/// as `speaker_names` gives.
fn translation_srt(transcript: &Transcript, speaker_names: HashMap<String, String>) -> String {
    translation_only(transcript).to_srt_with(
        SrtContent::Original,
        &SpeakerNames {
            text: speaker_names,
            ..SpeakerNames::default()
        },
    )
}

/// The Segment of `segments` with the times of `cue`, the one a translation's cue belongs to.
fn segment_at_times<'a>(segments: &'a [Segment], cue: &Segment) -> Option<&'a Segment> {
    segments
        .iter()
        .find(|segment| (segment.start_ms, segment.end_ms) == (cue.start_ms, cue.end_ms))
}

/// What a restore left behind: how many Segments it gave times that no translation lines up with.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize)]
pub struct Restoration {
    pub unmatched_count: usize,
}

impl Restoration {
    /// The Segments of `now` with times `previous` did not have, and no cue at them in one of
    /// `translations`; a translation lines up with its original by time alone.
    fn new(previous: &Transcript, now: &Transcript, translations: &[Transcript]) -> Restoration {
        let unmatched_count = now
            .segments
            .iter()
            .filter(|segment| segment_at_times(&previous.segments, segment).is_none())
            .filter(|segment| {
                translations
                    .iter()
                    .any(|translation| segment_at_times(&translation.segments, segment).is_none())
            })
            .count();
        Restoration { unmatched_count }
    }
}

/// The dialogue of a translation's cue, as written, for a Segment said by `speaker`. Tsuzuri
/// writes that Speaker's label before it, named as `names` gives, so that label is taken off, or
/// whatever label the line opens with when the name has changed since; a Segment with no Speaker
/// has no label to take off, so a line that only looks like one stays dialogue.
fn translated_dialogue(
    text: &str,
    speaker: Option<&str>,
    names: &HashMap<String, String>,
) -> String {
    let Some(speaker) = speaker else {
        return text.to_string();
    };
    let name = names.get(speaker).map_or(speaker, String::as_str);
    let dialogue = text
        .strip_prefix(name)
        .and_then(|rest| rest.strip_prefix([':', '：']))
        .map(|rest| rest.trim_start_matches([' ', '\t']))
        .or_else(|| match split_label(text) {
            (Some(_), dialogue) => Some(dialogue),
            (None, _) => None,
        });
    dialogue.unwrap_or(text).to_string()
}

/// `translation`'s cues, each with the Speaker of the Segment of `original` with its times in
/// place of the label it carried for that Segment of `previous`; a cue with other times belongs to
/// no Segment and stays as written.
fn translation_with_speakers(
    translation: &Transcript,
    original: &Transcript,
    previous: &Transcript,
    speaker_names: &HashMap<String, String>,
) -> Transcript {
    Transcript {
        segments: translation
            .segments
            .iter()
            .map(|cue| {
                let Some(segment) = segment_at_times(&original.segments, cue) else {
                    return cue.clone();
                };
                let before = segment_at_times(&previous.segments, cue)
                    .and_then(|segment| segment.speaker.as_deref());
                Segment {
                    speaker: segment.speaker.clone(),
                    text: translated_dialogue(&cue.text, before, speaker_names),
                    ..cue.clone()
                }
            })
            .collect(),
    }
}

/// Each translated Segment with its translation as its text and its text as its translation.
fn translation_first(transcript: &Transcript) -> Transcript {
    Transcript {
        segments: transcript
            .segments
            .iter()
            .map(|segment| match segment.translation.as_deref() {
                Some(translation) if !translation.trim().is_empty() => Segment {
                    text: translation.to_string(),
                    translation: Some(segment.text.clone()),
                    ..segment.clone()
                },
                _ => segment.clone(),
            })
            .collect(),
    }
}

/// Which text of a Segment an edit replaces.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SegmentField {
    Text,
    Translation,
    /// Written to the original and the translation shown; an empty one leaves the Segment with none.
    Speaker,
}

/// What a translation needs from the Project when it starts.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TranslationSource {
    pub generation: u64,
    pub directory: PathBuf,
    /// The Resource being translated.
    pub name: String,
    pub transcript: Transcript,
    /// The Primary Language it is translated from.
    pub language: Language,
    /// The Project Model to translate with in place of the general one.
    pub model: Option<PathBuf>,
}

/// What a transcription needs from the Project when it starts.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TranscriptionTarget {
    pub generation: u64,
    pub directory: PathBuf,
    /// The Resource being transcribed.
    pub name: String,
    pub media: PathBuf,
    /// Where the Resource's original subtitle is written.
    pub subtitle: PathBuf,
    pub language: Language,
    /// The Project Model to transcribe with in place of the general one.
    pub model: Option<PathBuf>,
    pub transcription: TranscriptionOverrides,
}

/// The files of a Project sharing one name.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Resource {
    pub name: String,
    pub media: Option<PathBuf>,
    /// The subtitle in the Primary Language.
    pub subtitle: Option<PathBuf>,
    /// A subtitle for each other Language, in the order of [`Language::ALL`].
    pub translations: Vec<(Language, PathBuf)>,
}

impl Resource {
    pub fn translation_path(&self, language: Language) -> Option<&Path> {
        self.translations
            .iter()
            .find(|(each, _)| *each == language)
            .map(|(_, path)| path.as_path())
    }
}

/// What a Project records about itself in its directory, so it opens the same way next time.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct ProjectConfig {
    /// The Primary Language, or none to follow the Interface Language.
    pub language: Option<Language>,
    /// The Language of the last translation.
    pub translation_language: Option<Language>,
    #[serde(flatten)]
    pub options: ProjectOptions,
}

/// What the user sets for one Project in the settings beside its Primary Language.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct ProjectOptions {
    pub bilingual_order: BilingualOrder,
    /// Whether each translation keeps a Bilingual SRT beside it, written with either of its texts.
    pub is_bilingual_autosaved: bool,
    /// Whether a subtitle about to be overwritten is first kept as a Backup.
    pub is_overwrite_backed_up: bool,
    pub models: ProjectModels,
    pub transcription: TranscriptionOverrides,
}

/// The Project Models by the slot each is chosen for; a slot without one uses the general settings' Model.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct ProjectModels {
    pub transcription: Option<PathBuf>,
    pub translation: Option<PathBuf>,
}

/// The Transcription Settings a Project sets for itself; one left `None` follows the general settings.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct TranscriptionOverrides {
    pub has_vad: Option<bool>,
    pub is_non_speech_suppressed: Option<bool>,
    pub is_context_carried: Option<bool>,
}

/// Which text a Bilingual SRT puts first in each cue and in its file name.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum BilingualOrder {
    #[default]
    OriginalFirst,
    TranslationFirst,
}

/// A Backup of one subtitle, by its file name in the history, the UTC time it was taken and its kind.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Backup {
    pub file: String,
    /// `YYYYMMDDTHHMMSSZ`
    pub taken_at: String,
    pub kind: BackupKind,
}

/// What a Backup keeps: what a Mode has just written, or a subtitle just before it was written over.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum BackupKind {
    Output,
    Overwrite,
}

File: src/backend/project.ts
import { invoke } from "@tauri-apps/api/core";
import { emit, listen, type UnlistenFn } from "@tauri-apps/api/event";

import type { TranscriptionSettings } from "./transcription";

export type { UnlistenFn };

export interface Segment {
  start_ms: number;
  end_ms: number;
  speaker?: string;
  text: string;
  translation?: string;
}

/** What the user sets for one Project in the settings beside its Primary Language. */
export interface ProjectOptions {
  bilingual_order: "original-first" | "translation-first";
  is_bilingual_autosaved: boolean;
  is_overwrite_backed_up: boolean;
  /** The Project Models; a slot without one uses the general settings' Model. */
  models: ProjectModels;
  transcription: TranscriptionOverrides;
}

export interface ProjectModels {
  transcription: string | null;
  translation: string | null;
}

/** The Transcription Settings a Project sets for itself; `null` follows the general ones. */
export type TranscriptionOverrides = {
  [Setting in keyof TranscriptionSettings]: boolean | null;
};

/** A Resource as the Resource list shows it. */
export interface ResourceView {
  name: string;
  has_media: boolean;
  has_subtitle: boolean;
  /** The Language codes of its translation files. */
  translation_languages: string[];
}

/** The Project as Rust holds it; the webview only ever shows this, never a copy of its own. */
export interface ProjectView {
  directory: string;
  /** The Primary Language code. */
  language: string;
  /** The Language code of the last translation. */
  translation_language: string | null;
  options: ProjectOptions;
  translation_glossary: TranslationGlossaryView | null;
  resources: ResourceView[];
  current_resource: string | null;
  /** The Current Resource's media file. */
  media: string | null;
  /** The Current Resource's Segments. */
  segments: Segment[];
  /** The Language code of the translations the Segments carry. */
  shown_translation: string | null;
  /** Whether the Current Resource has a change to undo. */
  has_undo: boolean;
  /** Whether the Current Resource has an undone change to redo. */
  has_redo: boolean;
  /** The Mode running on the Current Resource, holding the subtitles it writes. */
  running_mode: RunningMode | null;
  /** The Segments the running translation works on now, by position. */
  pending_batch: SegmentSpan | null;
}

/** A transcription holds every subtitle of its Resource; a translation, the one it writes. */
/** The Segments from `first` through `last`, by position. */
export interface SegmentSpan {
  first: number;
  last: number;
}

export type RunningMode =
  { mode: "transcription" } | { mode: "translation"; language: string };

/** The Current Resource as the Resource list shows it, or none. */
export function currentResource(
  project: ProjectView | null,
): ResourceView | undefined {
  return project?.resources.find(
    (resource) => resource.name === project.current_resource,
  );
}

/** The file a Translation Glossary came from, how many terms it holds, and the Speakers it names in the Primary Language. */
export interface TranslationGlossaryView {
  file: string;
  term_count: number;
  speakers: string[];
}

/** The Project Rust holds now, or none before one is made. */
export function currentProject(): Promise<ProjectView | null> {
  return invoke<ProjectView | null>("current_project");
}

/**
 * The Project Rust holds, read once for each change and handed to every follower in the order they
 * began to follow, then to each `afterEach` callback. A read answered after a later one is dropped,
 * so no follower is shown an older Project than it already has.
 */
export class ProjectFeed {
  /** The Project last read, or `undefined` before the first read. */
  private latest: ProjectView | null | undefined = undefined;
  private readonly followers = new Set<(project: ProjectView | null) => void>();
  private readonly settlers: (() => void)[] = [];
  private asked = 0;
  private shown = 0;

  /** The Project last read, or none. */
  get project(): ProjectView | null {
    return this.latest ?? null;
  }

  /** Calls `show` with the Project read last, if any, and again with each one read after it. */
  follow(show: (project: ProjectView | null) => void): () => void {
    this.followers.add(show);
    if (this.latest !== undefined) show(this.latest);
    return () => this.followers.delete(show);
  }

  /** Calls `settle` once every follower has been shown a Project. */
  afterEach(settle: () => void): void {
    this.settlers.push(settle);
  }

  /** Reads the Project now and each time Rust announces a change. */
  async start(): Promise<UnlistenFn> {
    const unlisten = await listen("project-changed", () => void this.read());
    await this.read();
    return unlisten;
  }

  private async read(): Promise<void> {
    const asked = ++this.asked;
    const project = await currentProject();
    if (asked < this.shown) return;
    this.shown = asked;
    this.latest = project;
    for (const show of this.followers) show(project);
    for (const settle of this.settlers) settle();
  }
}

/** Undo or Redo chosen from the Edit menu. */
export type EditCommand = "undo" | "redo";

/** Calls `apply` with each Undo or Redo chosen from the Edit menu, which takes their shortcuts first. */
export function followEditCommands(
  apply: (command: EditCommand) => void,
): Promise<UnlistenFn> {
  return listen<EditCommand>("edit-command", (event) => apply(event.payload));
}

/** Asks every view to read the Project again, as when Rust announces nothing after a refusal. */
export function refreshProject(): Promise<void> {
  return emit("project-changed");
}

/** Opens `path` as the Project, a directory or the directory of an SRT file, in `language` when the directory records none. */
export function openProject(
  command: "open_project" | "open_srt",
  path: string,
  language: string,
): Promise<void> {
  return invoke(command, { path, language });
}

export function selectResource(name: string | undefined): Promise<void> {
  return invoke("select_resource", { name });
}

export function setPrimaryLanguage(language: string): Promise<void> {
  return invoke("set_primary_language", { language });
}

export function setProjectOptions(options: ProjectOptions): Promise<void> {
  return invoke("set_project_options", { options });
}

/** Pairs the Project's files again and reads the Current Resource again from them. */
export function reloadProject(): Promise<void> {
  return invoke("reload_project");
}

export function showTranslation(language: string | null): Promise<void> {
  return invoke("show_translation", { language });
}

/** Which texts an SRT written from the Current Resource carries. */
export type SrtContent = "original" | "translation" | "bilingual";

export function exportPath(content: SrtContent): Promise<string> {
  return invoke<string>("export_path", { content });
}

export function saveSrt(path: string, content: SrtContent): Promise<void> {
  return invoke("save_srt", { path, content });
}

/** A Backup as Rust lists it: its file name in the history, the UTC time it was taken and its kind. */
export interface Backup {
  file: string;
  /** `YYYYMMDDTHHMMSSZ` */
  taken_at: string;
  kind: BackupKind;
}

/** What a Mode has just written, or a subtitle just before it was written over. */
export type BackupKind = "output" | "overwrite";

/** The Backups of the original, with no Language, or of one translation. */
export interface SubtitleVersions {
  language: string | null;
  backups: Backup[];
}

/** One cue of a Version as a Comparison Row shows it. */
export interface ComparedCue {
  start_ms: number;
  end_ms: number;
  text: string;
}

/** How the cues of a Comparison Row stand to each other. */
export type RowKind = "pair" | "addition" | "removal" | "split" | "merge";

/** The cues of two Versions that cover the same speech, and what changed between them. */
export interface ComparedRow {
  kind: RowKind;
  left: ComparedCue[];
  right: ComparedCue[];
  is_text_changed: boolean;
  is_time_changed: boolean;
  /** A Pair's changed text character by character; empty otherwise. */
  text_spans: TextSpan[];
}

/** A run of characters of a Pair's text: in both Versions, only the earlier, or only the later. */
export interface TextSpan {
  kind: "common" | "removal" | "addition";
  text: string;
}

export function subtitleVersions(): Promise<SubtitleVersions[]> {
  return invoke<SubtitleVersions[]>("subtitle_versions");
}

/** The cues of two Versions of one subtitle side by side, where none names the subtitle as it is now. */
export function compareVersions(
  language: string | null,
  left: string | null,
  right: string | null,
): Promise<ComparedRow[]> {
  return invoke<ComparedRow[]>("compare_versions", { language, left, right });
}

/** What a restore left behind: how many Segments it gave times that no translation lines up with. */
export interface Restoration {
  unmatched_count: number;
}

export function restoreVersion(
  language: string | null,
  backup: string | undefined,
): Promise<Restoration> {
  return invoke<Restoration>("restore_version", { language, backup });
}

/** The cues of the Current Resource's translation into `language`, as its file is written. */
export function translationCues(language: string): Promise<ComparedCue[]> {
  return invoke<ComparedCue[]>("translation_cues", { language });
}

/** Which part of a Comparison Row to take back. */
export type RevertPart = "text" | "times" | "whole";

/** Takes back one Comparison Row of `backup` against the subtitle in `language`, as they compare now. */
export function revertRow(
  language: string | null,
  backup: string,
  row: number,
  part: RevertPart,
): Promise<Restoration> {
  return invoke<Restoration>("revert_row", { language, backup, row, part });
}

/** One term: its word in each Language, and whether it names a Speaker. */
export interface GlossaryRow {
  words: string[];
  is_speaker: boolean;
}

/** The Translation Glossary laid out for editing, named as Rust names it. */
export interface GlossaryTable {
  languages: string[];
  rows: GlossaryRow[];
  has_source_target_header: boolean;
}

export function translationGlossaryTable(): Promise<GlossaryTable> {
  return invoke<GlossaryTable>("translation_glossary_table");
}

export function saveTranslationGlossary(rows: GlossaryRow[]): Promise<void> {
  return invoke("save_translation_glossary", { rows });
}
